geo/src/algorithm/centroid.rs

```diff
@@ -335,42 +163,324 @@ where
+        }
+    }
+
+    fn add_multi_line_string(&mut self, multi_line_string: &MultiLineString<T>) {
+        if self.centroid_dimensions() > OneDimensional {
+            return;
+        }
+
+        for element in &multi_line_string.0 {
+            self.add_line_string(element);
+        }
+    }
+
+    fn add_polygon(&mut self, polygon: &Polygon<T>) {
+        // Polygons which are completely covered by their interior rings have zero area, and
+        // represent a unique degeneracy into a line_string which cannot be handled by accumulating
+        // directly into `self`. Instead, we perform a sub-operation, inspect the result, and only
+        // then incorporate the result into `self.
+
+        let mut exterior_operation = CentroidOperation::new();
+        exterior_operation.add_ring(polygon.exterior());
+
+        let mut interior_operation = CentroidOperation::new();
+        for interior in polygon.interiors() {
+            interior_operation.add_ring(interior);
+        }
+
+        if let Some(exterior_weighted_centroid) = exterior_operation.0 {
+            let mut poly_weighted_centroid = exterior_weighted_centroid;
+            if let Some(interior_weighted_centroid) = interior_operation.0 {
+                poly_weighted_centroid.sub_assign(interior_weighted_centroid);
+                if poly_weighted_centroid.weight.is_zero() {
+                    // A polygon with no area `interiors` completely covers `exterior`, degenerating to a linestring
+                    self.add_line_string(polygon.exterior());
+                    return;
+                }
+            }
+            self.add_weighted_centroid(poly_weighted_centroid);
+        }
+    }
+
+    fn add_multi_point(&mut self, multi_point: &MultiPoint<T>) {
+        if self.centroid_dimensions() > ZeroDimensional {
+            return;
+        }
+
+        for element in &multi_point.0 {
+            self.add_coord(element.0);
+        }
+    }
+
+    fn add_multi_polygon(&mut self, multi_polygon: &MultiPolygon<T>) {
+        for element in &multi_polygon.0 {
+            self.add_polygon(element);
+        }
+    }
+
+    fn add_geometry_collection(&mut self, geometry_collection: &GeometryCollection<T>) {
+        for element in &geometry_collection.0 {
+            self.add_geometry(element);
+        }
+    }
+
+    fn add_rect(&mut self, rect: &Rect<T>) {
+        match rect.dimensions() {
+            ZeroDimensional => self.add_coord(rect.min()),
+            OneDimensional => self.add_line(&Line::new(rect.min(), rect.max())),
+            TwoDimensional => {
+                self.add_centroid(TwoDimensional, rect.centroid().0, rect.unsigned_area())
+            }
+            Empty => unreachable!("Rect dimensions cannot be empty"),
+        }
+    }
+
+    fn add_triangle(&mut self, triangle: &Triangle<T>) {
+        match triangle.dimensions() {
+            ZeroDimensional => self.add_coord(triangle.0),
+            OneDimensional => {
```

Comment:
This seems different from how we are handling a degenerate ring.  I think we should simply add all three line segments in this case rather than just the longest one.  The centroid point itself doesn't change, but the weight would be double what it is now.  

These sort of degenerate cases are not really valid as per SFS, so we just have to be internally consistent.  Below is a natural test case that users would expect to pass (IMO).

``` rust
#[test]
fn degenerate_triangle_like_ring() {
    let triangle = Triangle(c(0., 0.), c(1., 1.), c(2., 2.));
    let poly: Polygon<_> = triangle.clone().into();

    let line = Line::new(c(0., 1.), c(1., 3.));

    let g1 = GeometryCollection(vec![triangle.into(), line.into()]);
    let g2 = GeometryCollection(vec![poly.into(), line.into()]);
    assert_eq!(g1.centroid(), g2.centroid());
}
```

Reply:
Yeah, that's a great point. 

Plus your suggestion is simpler. =)

I've rebased and pushed up one new commit to address this for Triangles and Rects (which had the same issue).